sdk/nodejs/runtime/invoke.ts

```diff
@@ -12,31 +12,67 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+import * as fs from "fs";
 import * as grpc from "grpc";
+import * as asset from "../asset";
 import { InvokeOptions } from "../invoke";
 import * as log from "../log";
-import { Inputs } from "../output";
+import { Inputs, Output } from "../output";
 import { debuggablePromise } from "./debuggable";
 import { deserializeProperties, serializeProperties, unknownValue } from "./rpc";
-import { excessiveDebugOutput, getMonitor, rpcKeepAlive } from "./settings";
+import { excessiveDebugOutput, getMonitor, getSyncInvokes, rpcKeepAlive } from "./settings";
+
+import { ProviderRef, Resource } from "../resource";
+import * as utils from "../utils";
 
 const gstruct = require("google-protobuf/google/protobuf/struct_pb.js");
 const providerproto = require("../proto/provider_pb.js");
 
 /**
- * invoke dynamically invokes the function, tok, which is offered by a provider plugin.  The inputs
- * can be a bag of computed values (Ts or Promise<T>s), and the result is a Promise<any> that
- * resolves when the invoke finishes.
+ * `invoke` dynamically invokes the function, `tok`, which is offered by a provider plugin. `invoke`
+ * behaves differently in the case that options contains `{async:true}` or not.
+ *
+ * In the case where `{async:true}` is present in the options bag:
+ *
+ * 1. the result of `invoke` will be a Promise resolved to the result value of the provider plugin.
+ * 2. the `props` inputs can be a bag of computed values (Ts or Promise<T>s).
+ *
+ *
+ * In the case where `{async:true}` is not present in the options bag:
+ *
+ * 1. the result of `invoke` will be a Promise resolved to the result value of the provider call.
+ *    However, that Promise will *also* have the respective values of the Provider result exposed
+ *    directly on it as properties.
```

Comment:
This is to preserve compat with existing libraries that call into the new pulumi/pulumi.  I want to preserve the behavior that you get a Promise back with properties lifted to it.    Importantly, the promise we return involves no actual asynchrony.  We do everything synchronously, but just return `Promise.resolve(syncValue)` back.